### PR TITLE
Helm chart improvements and added ovms operator

### PR DESCRIPTION
**Pull request by dtrawins**

### CODE REVIEW

File: deploy/README.md [hunk 1]
```diff
@@ -27,46 +27,108 @@ You can download the model from the link provided above and upload it to GCS:
 ```shell script
 $ gsutil cp -r 1 gs://model-repository/1
 ```
+The models repository can be also distributed on the cluster nodes in the local path or it could be stored on the Kubernetes persistent volume.
+Below are described supported storage options:
```
Comment (rasapala) [inline deploy/README.md:31]: The supported storage options are described below:

File: deploy/README.md [hunk 2]
```diff
@@ -27,46 +27,108 @@ You can download the model from the link provided above and upload it to GCS:
...
+list permission on the model version folders.
+
+### Persistent Volume
+If is possible to deploy OVMS using Kubernetes [persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
+
+That opens a possibility of storing the using the models for OVMS on all Kubernetes [supported filesystems](https://kubernetes.io/docs/concepts/storage/storage-classes/).
+
+In the helm set the parameter `models_volume_claim` with the name of the `PersistentVolumeClaim` record with the models. While set, it will be mounted as `/models` folder inside the OVMS container.
+
+Note that parameter `models_volume_claim` is excluding with `models_host_path`. Only one of them should be set.
```
Comment (rasapala) [inline deploy/README.md:86]: is mutually exclusive with

File: deploy/README.md [hunk 3]
```diff
@@ -27,46 +27,108 @@ You can download the model from the link provided above and upload it to GCS:
...
+
+## Assigning Resource Specs
+
+You can restrict assigned cluster resources to the OVMS container but setting the parameter `resources`.
+Be default, there are no restrictions but it could be used to reduce the CPU and memory like in the snippet example from the values.yaml file:
```
Comment (rasapala) [inline deploy/README.md:91]: By

File: extras/ovms-operator/Makefile [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+IMG ?= controller:latest
+
+all: docker-build
+
+# Run against the configured Kubernetes cluster in ~/.kube/config
+run: helm-operator
+	$(HELM_OPERATOR) run
+
+# Install CRDs into a cluster
+install: kustomize
```
Comment (rasapala) [inline extras/ovms-operator/Makefile:24]: customize? here and below. kustomize -> customize

**Review by rasapala — CHANGES_REQUESTED** [review]

File: extras/ovms-operator/Makefile (hunk 4 above)
Comment (dtrawins) [reply extras/ovms-operator/Makefile:24]: that is correct name. it comes from https://kustomize.io/ tool for automating deployments in k8s

**Review by rasapala — APPROVED** [review]
LGTM

File: deploy/README.md [hunk 5]
```diff
@@ -10,169 +10,259 @@ inference requests to the running server.
...
+Use OVMS with models stored on azure blob storage by providing `azure_storage_connection_string` parameter. Model path should follow `az` scheme like below:
+```shell script
+helm install ovms-app ovms --set model_name=resnet,model_path=az://bucket/model_path,azure_storage_connection_string="DefaultEndpointsProtocol=https;AccountName=azure_account_name;AccountKey=smp/hashkey==;EndpointSuffix=core.windows.net"
+```
+ 
+### Local Node Storage
+Beside the cloud storage, models could be stored locally on the kubernetes nodes filesystem.
+Use the parameter `models_host_path` with the local path on the nodes. It will be mounted in the OVMS container as `/models` folder.
+
+While the models folder is mounted in the OVMS container, the parameter `model_path` should refer to the path staring with /models/... and point to the folder with the model versions.
```
Comment (mzegla) [inline deploy/README.md:73]:
```suggestion
While the models folder is mounted in the OVMS container, the parameter `model_path` should refer to the path starting with /models/... and point to the folder with the model versions.
```

File: deploy/README.md [hunk 6]
```diff
@@ -10,169 +10,259 @@ inference requests to the running server.
...
+
+Note that the OVMS container starts, by default, with the security context of account `ovms` with pid 5000 ahd group 5000.
```
Comment (mzegla) [inline deploy/README.md:75]:
```suggestion
Note that the OVMS container starts, by default, with the security context of account `ovms` with pid 5000 and group 5000. 
```

File: deploy/README.md [hunk 7]
```diff
@@ -10,169 +10,259 @@ inference requests to the running server.
...
+Use the parameter `models_host_path` with the local path on the nodes. It will be mounted in the OVMS container as `/models` folder.
+
+While the models folder is mounted in the OVMS container, the parameter `model_path` should refer to the path staring with /models/... and point to the folder with the model versions.
+
+Note that the OVMS container starts, by default, with the security context of account `ovms` with pid 5000 ahd group 5000. 
+If the mounted models have restricted access permissions, change the security context of the OVMS service or adjust permissions to the models. OVMS requires read permissions on the model files and 
+list permission on the model version folders.
+
+### Persistent Volume
+If is possible to deploy OVMS using Kubernetes [persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
```
Comment (mzegla) [inline deploy/README.md:80]:
```suggestion
It is possible to deploy OVMS using Kubernetes [persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
```

File: deploy/README.md [hunk 8]
```diff
@@ -10,169 +10,259 @@ inference requests to the running server.
...
+
+That opens a possibility of storing the using the models for OVMS on all Kubernetes [supported filesystems](https://kubernetes.io/docs/concepts/storage/storage-classes/).
```
Comment (mzegla) [inline deploy/README.md:82]:
```suggestion
That opens a possibility of storing the models for OVMS on all Kubernetes [supported filesystems](https://kubernetes.io/docs/concepts/storage/storage-classes/).
```

File: deploy/README.md [hunk 9]
```diff
@@ -10,169 +10,259 @@ inference requests to the running server.
...
+
+In the helm set the parameter `models_volume_claim` with the name of the `PersistentVolumeClaim` record with the models. While set, it will be mounted as `/models` folder inside the OVMS container.
+
+Note that parameter `models_volume_claim` is mutually exclusive with `models_host_path`. Only one of them should be set.
+
+## Assigning Resource Specs
+
+You can restrict assigned cluster resources to the OVMS container but setting the parameter `resources`.
```
Comment (mzegla) [inline deploy/README.md:90]:
```suggestion
You can restrict assigned cluster resources to the OVMS container by setting the parameter `resources`.
```

File: deploy/README.md [hunk 10]
```diff
@@ -10,169 +10,259 @@ inference requests to the running server.
...
+By default, there are no restrictions but that parameter could be used to reduce the CPU and memory allocation. Below is the snippet example from the values.yaml file:
+```yaml
+resources:
+  limits:
+    cpu: 8.0
+    memory: 512Mi
+```
+Beside setting the CPU and memory resources, the same parameter can be used to assign AI accelerators like iGPU, or VPU.
+That assume using adequate Kubernetes device plugin from [Intel Device Plugin for Kubernetes](https://github.com/intel/intel-device-plugins-for-kubernetes).
```
Comment (mzegla) [inline deploy/README.md:99]:
```suggestion
That assumes using adequate Kubernetes device plugin from [Intel Device Plugin for Kubernetes](https://github.com/intel/intel-device-plugins-for-kubernetes).
```

File: deploy/ovms/templates/aws_secret.yaml [hunk 11]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2020 Intel Corporation
+# Copyright (c) 2020-2001 Intel Corporation
```
Comment (mzegla) [inline deploy/ovms/templates/aws_secret.yaml:2]:
```suggestion
# Copyright (c) 2020-2021 Intel Corporation
```

File: deploy/ovms/templates/deployment.yaml [hunk 12]
```diff
@@ -42,84 +53,113 @@ spec:
...
+        {{- if .Values.config_configmap_name }}
         - name: config
           mountPath: "/config"
           readOnly: true
-        {{ end }}
-      {{ if or .Values.gcp_creds_secret_name .Values.config_configmap_name }}
+        {{- end }}
+        {{- if or .Values.models_host_path .Values.models_volume_claim }}
+        - name: models
+          mountPath: "/models"
```
Comment (mzegla) [inline deploy/ovms/templates/deployment.yaml:137]: don't we want to parametrize this path?

File: deploy/ovms/values.yaml [hunk 13]
```diff
@@ -26,11 +30,26 @@ plugin_config: >-
...
 service_type: "LoadBalancer"
+# alternatively set to NodePort or ClusterIP
+resources: {}
+  #  limits:
+  #    cpu: 8.0
+  #    memory: 512Mi
+  #    gpu.intel.com/i915: 1
+security_context: {}
+  #  runAsUser: 5000
+  #  runAsGroup: 5000
```
Comment (mzegla) [inline deploy/ovms/values.yaml:49]: Are commented values some kind of example or did you miss removing them after development?

File: extras/ovms-operator/README.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+It allows for easy deployment and management of OVMS service in the Kubernetes cluster by just creating `Ovms` resource
+record.
+```bash
+kubectl apply -f config/samples/intel_v1alpha1_ovms.yaml
+```
+
+## Operator deployment
+Deploy the operator using the steps covered in [OperatorHub](https://operatorhub.io) (TBD)
+Alternatively you are are not using OLM component, run commands:
```
Comment (mzegla) [inline extras/ovms-operator/README.md:12]:
```suggestion
Alternatively, if you are not using OLM component, run commands:
```

File: extras/ovms-operator/config/samples/intel_v1alpha1_ovms.yaml [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  gcp_creds_secret_name: ""
+  grpc_port: 8080
+  image_name: openvino/model_server:latest
+  log_level: INFO
+  model_name: "resnet"
+  model_path: "gs://ovms-public-eu/resnet50-binary"
+  models_host_path: ""
+  models_volume_claim: ""
+  nireq: ""
+  plugin_config: '{\"CPU_THROUGHPUT_STREAMS\":\"1"}'
```
Comment (mzegla) [inline extras/ovms-operator/config/samples/intel_v1alpha1_ovms.yaml:20]: Didn't you miss one escape?
```suggestion
  plugin_config: '{\"CPU_THROUGHPUT_STREAMS\":\"1\"}'
```

File: extras/ovms-operator/helm-charts/ovms/values.yaml [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+log_level: INFO
+service_type: "LoadBalancer"
+resources: {}
+  #  limits:
+  #    cpu: 8.0
+  #    memory: 512Mi
+  #    gpu.intel.com/i915: 1
+security_context: {}
+  #  runAsUser: 5000
+  #  runAsGroup: 5000
```
Comment (mzegla) [inline extras/ovms-operator/helm-charts/ovms/values.yaml:46]: Are those comments left here on purpose?

File: extras/ovms-operator/tests/README.md [hunk 17]
```diff
@@ -0,0 +1,72 @@
+# Operator testing procedure
+
+## Building the operator image 
+
+```
+export REGISTRY=ger-registry-pre.caas.intel.com/ovms
```
Comment (mzegla) [inline extras/ovms-operator/tests/README.md:6]: is that registry open to the public?

**Review by mzegla — COMMENTED** [review]
It looks good, but I saw that in multiple places grpc and rest ports are not parametrized. 
They are hard coded to 8080 and 8081. Could you check that? Perhaps we should change it.

File: deploy/ovms/values.yaml (hunk 13 above)
Comment (dtrawins) [reply deploy/ovms/values.yaml:49]: they are examples

**Review by mzegla — APPROVED** [review]

**Pull request merged**